server/chat-plugins/chatlog.ts

```diff
@@ -626,6 +663,176 @@ export abstract class Searcher {
+		const stats: RoomStats[] = [];
+		for (const day of days) {
+			const curStats = await this.dayStats(room, day);
+			if (!curStats) continue;
+			stats.push(curStats);
+		}
+		// now, having collected the stats for each day, we need to merge them together
+		const collected: RoomStats = {
+			deadTime: 0,
+			deadPercent: 0,
+			lines: {},
+			users: {},
+			days: days.length,
+			linesPerUser: 0,
+			totalLines: 0,
+			averagePresent: 0,
+		};
+
+		// merge
+		for (const entry of stats) {
+			for (const k of ['deadTime', 'deadPercent', 'linesPerUser', 'totalLines', 'averagePresent'] as const) {
+				collected[k] += entry[k];
+			}
+			for (const type of ['lines'] as const) {
+				for (const k in entry[type]) {
+					if (!collected[type][k]) collected[type][k] = 0;
+					collected[type][k] += entry[type][k];
+				}
+			}
+		}
+
+		// average
+		for (const k of ['deadTime', 'deadPercent', 'linesPerUser', 'totalLines', 'averagePresent'] as const) {
+			collected[k] /= stats.length;
+		}
+
+		return {average: collected, days: stats};
+	}
+	async dayStats(room: RoomID, day: string) {
+		const cached = this.roomstatsCache.get(day);
+		if (cached) return cached;
+		const results: RoomStats & {day: string} = {
+			deadTime: 0,
+			deadPercent: 0,
+			lines: {},
+			users: {},
+			days: 1, // irrelevant
+			linesPerUser: 0,
+			totalLines: 0,
+			averagePresent: 0,
+			day,
+		};
+		const path = FS(`logs/chat/${room}/${LogReader.getMonth(day)}/${day}.txt`);
+		if (!path.existsSync()) return false;
+		const stream = path.createReadStream();
+		let lastTime = new Date(day).getTime(); // start at beginning of day to be sure
+		let userstatCount = 0;
+		const waitIncrements = [];
+		for await (const line of stream.byLine()) {
+			const [, type, ...rest] = line.split('|');
+			switch (type) {
+			// the actual info in this is unused, but it may be useful in the future (we use the keys later)
+			case 'J': case 'j': {
+				if (rest[0]?.startsWith('*')) continue; // ignore bots
+				const userid = toID(rest[0]);
+				if (!results.users[userid]) {
+					results.users[userid] = 0;
+				}
+				results.users[userid]++;
+				break;
+			}
+			case 'c:': case 'c': {
+				const {time, username} = LogViewer.parseChatLine(line, day);
+				const curTime = time.getTime();
+				if (curTime - lastTime > 5 * 60 * 1000) { // more than 5 minutes
+					waitIncrements.push(curTime - lastTime);
+					lastTime = curTime;
+				}
+				const userid = toID(username);
+				if (!results.lines[userid]) results.lines[userid] = 0;
+				results.lines[userid]++;
+				results.totalLines++;
+				break;
+			}
+			case 'userstats': {
+				const [rawTotal] = rest;
+				const total = parseInt(rawTotal.split(':')[1]);
+				results.averagePresent += total;
+				userstatCount++;
+				break;
+			}
+			}
+		}
+		results.deadTime = waitIncrements.length ? waitIncrements.reduce((c, p) => c + p) / waitIncrements.length : 0;
```

Comment:
Can this just be a for-loop? I'm not a fan of `reduce` for things like this.

Reply:
Done.